File: javascript/08-stuttering-classification.js
/**
 * Stuttering Classification — Vocametrix API example (Node.js)
 *
 * Classifies stuttering patterns in speech. Async endpoint:
 *   POST /api/classify-stuttering  → session_id
 *   Poll GET /api/therapy-status/:id  until complete
 *   Fetch GET /api/therapy-result/:id
 *
 * Server timeout: 600 seconds.
 * The result has a `classification` array of ~4s blocks, each with a stutter
 * `primaryType`, a `transcription`, and a `words` array of per-word timestamps.
 *
 * Upload pattern: POST /api/assignFileId → POST /api/classify-stuttering
 * Auth: X-API-Key header
 *
 * Usage:
 *   node 08-stuttering-classification.js path/to/audio.wav
 */

import fs from 'fs';
import FormData from 'form-data';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';
const HEADERS = { 'X-API-Key': API_KEY };
const POLL_INTERVAL_MS = 5000;
const MAX_WAIT_MS = 620_000;

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function assignFileId(audioPath) {
  const form = new FormData();
  form.append('audio', fs.createReadStream(audioPath));
  const res = await fetch(`${BASE_URL}/api/assignFileId`, {
    method: 'POST',
    headers: { ...HEADERS, ...form.getHeaders() },
    body: form,
  });
  if (!res.ok) throw new Error(`assignFileId failed: ${await res.text()}`);
  return (await res.json()).fileId;
}

async function classifyStuttering(audioPath) {
  const fileId = await assignFileId(audioPath);

  const startRes = await fetch(`${BASE_URL}/api/classify-stuttering`, {
    method: 'POST',
    headers: { ...HEADERS, 'Content-Type': 'application/json' },
    // Optional flags: transcribe (default true → adds per-word timestamps),
    // includePhonemes (default false → set true for per-block phonetic output).
    body: JSON.stringify({ fileId }),
  });
  if (!startRes.ok) throw new Error(`classify-stuttering failed: ${await startRes.text()}`);
  const { session_id } = await startRes.json();
  console.log(`Classification started: session_id=${session_id}`);

  let elapsed = 0;
  while (elapsed < MAX_WAIT_MS) {
    await sleep(POLL_INTERVAL_MS);
    elapsed += POLL_INTERVAL_MS;

    const statusRes = await fetch(`${BASE_URL}/api/therapy-status/${session_id}`, { headers: HEADERS });
    statusRes.raise_for_status?.();
    const status = await statusRes.json();
    const state = status.status ?? status.state ?? '';
    console.log(`  [${Math.round(elapsed / 1000)}s] ${state} ${status.progress_percent ?? status.progress ?? ''}`);

    if (state === 'complete' || status.result_available) break;
    if (['failed', 'error'].includes(state)) throw new Error(`Classification failed: ${JSON.stringify(status)}`);
  }

  const resultRes = await fetch(`${BASE_URL}/api/therapy-result/${session_id}`, { headers: HEADERS });
  if (!resultRes.ok) throw new Error(`therapy-result failed: ${await resultRes.text()}`);
  return resultRes.json();
}

function printResults(results) {
  console.log('\n=== Stuttering Classification Results ===');

  const overall = results.overallClassification ?? {};
  if (overall.primary_type) {
    console.log(`Overall: ${overall.primary_type} (${overall.severity ?? '?'}) - `
      + `${overall.stuttering_percentage ?? '?'}% of speech stuttered`);
  }

  const blocks = results.classification ?? [];
  console.log(`\nBlocks (${blocks.length}):`);
  for (const b of blocks) {
    console.log(`\n  [${b.startTime}s - ${b.stopTime}s]  ${b.primaryType}  |  ${JSON.stringify(b.transcription)}`);
    // Per-word timestamps (present when transcribe=true, the default).
    for (const w of b.words ?? []) {
      console.log(`        ${w.start.toFixed(2)}s - ${w.end.toFixed(2)}s   ${w.word}`);
    }
  }
}

const audioPath = process.argv[2];
if (!audioPath) {
  console.error('Usage: node 08-stuttering-classification.js <audio.wav>');
  process.exit(1);
}
classifyStuttering(audioPath).then(printResults).catch(err => { console.error(err.message); process.exit(1); });

File: python/08_stuttering_classification.py
"""
Stuttering Classification — Vocametrix API example

Classifies stuttering patterns in speech audio. This is an async endpoint:
  1. POST /api/classify-stuttering  → returns session_id
  2. Poll GET /api/therapy-status/:session_id  until complete
  3. Fetch GET /api/therapy-result/:session_id

The server timeout is 600 seconds (CLASSIFY_TIMEOUT_MS env var).
The result has a ``classification`` array of ~4s blocks, each with a stutter
``primaryType``, a ``transcription``, and a ``words`` array of per-word timestamps.

Upload pattern: POST /api/assignFileId → POST /api/classify-stuttering (async)
Auth: X-API-Key header on all endpoints

Usage:
    python 08_stuttering_classification.py path/to/audio.wav
"""

import os
import time
import argparse
import requests
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}

POLL_INTERVAL = 5   # seconds between status checks
MAX_WAIT = 620      # slightly above server timeout


def assign_file_id(audio_path: str) -> str:
    with open(audio_path, 'rb') as f:
        r = requests.post(
            f'{BASE_URL}/api/assignFileId',
            headers=HEADERS,
            files={'audio': f},
        )
    r.raise_for_status()
    return r.json()['fileId']


def classify_stuttering(audio_path: str) -> dict:
    file_id = assign_file_id(audio_path)

    r = requests.post(
        f'{BASE_URL}/api/classify-stuttering',
        headers=HEADERS,
        # Optional flags: transcribe (default True -> adds per-word timestamps),
        # includePhonemes (default False -> set True for per-block phonetic output).
        json={'fileId': file_id},
    )
    r.raise_for_status()
    session_id = r.json()['session_id']
    print(f'Classification started: session_id={session_id}')

    # Poll for completion
    elapsed = 0
    while elapsed < MAX_WAIT:
        time.sleep(POLL_INTERVAL)
        elapsed += POLL_INTERVAL

        status_r = requests.get(
            f'{BASE_URL}/api/therapy-status/{session_id}',
            headers=HEADERS,
        )
        status_r.raise_for_status()
        status = status_r.json()
        state = status.get('status', status.get('state', ''))
        progress = status.get('progress_percent', status.get('progress', ''))
        print(f'  [{elapsed:3d}s] {state} {progress}')

        if state == 'complete' or status.get('result_available'):
            break
        if state in ('failed', 'error'):
            raise RuntimeError(f'Classification failed: {status}')

    result_r = requests.get(
        f'{BASE_URL}/api/therapy-result/{session_id}',
        headers=HEADERS,
    )
    result_r.raise_for_status()
    return result_r.json()


def print_results(results: dict):
    print('\n=== Stuttering Classification Results ===')

    overall = results.get('overallClassification', {})
    if overall.get('primary_type'):
        print(f"Overall: {overall['primary_type']} ({overall.get('severity', '?')}) "
              f"- {overall.get('stuttering_percentage', '?')}% of speech stuttered")

    blocks = results.get('classification', [])
    print(f'\nBlocks ({len(blocks)}):')
    for b in blocks:
        print(f"\n  [{b.get('startTime')}s - {b.get('stopTime')}s]  "
              f"{b.get('primaryType')}  |  {b.get('transcription')!r}")
        # Per-word timestamps (present when transcribe=True, the default).
        for w in b.get('words', []):
            print(f"        {w['start']:.2f}s - {w['end']:.2f}s   {w['word']}")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Stuttering classification')
    parser.add_argument('audio', help='Path to WAV file (connected speech)')
    args = parser.parse_args()

    results = classify_stuttering(args.audio)
    print_results(results)
